File: scripts/vcf2gl.py
#! /usr/bin/python


# This scripts converts a vcf file to a simpler format for downstream analyses.
# Zach Gompert is calling this format genotype likelihood (gl). The first line
# lists the number of individuals and loci. The next line has individual ids.
# This is followed by one line per SNP that gives the SNP id (scaffold,
# position) and the phred-scaled genotype likelihoods, three per individual.
# A separate file with the allele frequency for each locus is being written.
# This version does not yet include a filter for maf.
#
# Usage: ./vcf2gl.py filtered_vcf_file.vcf ref_allele_freqs.txt alt_allele_freqs.txt > outfile.gl


import sys
import re
from collections import OrderedDict


rf_file = open(sys.argv[2], 'a')
af_file = open(sys.argv[3], 'a')

with open(sys.argv[1], 'rb') as file:
    n_snv = 0
    n_no_num = 0
    n_num = 0
    line_dict = dict()
    for line in file:
        if line[0:2] == '##':
            continue
        elif line[0:2] == '#C':
            header = line.split('\n')[0].split('\t')
            ind_ids = header[9:len(header)]
            #print line, header, ind_ids
        else:
            line_list = line.split('\t')
            if len(line_list[4]) > 1:
                continue
            else:
                geno_likely = OrderedDict()
                n_snv += 1
                n_ind = len(ind_ids)
                scaffold = line_list[0].split('d')[1]   # if "ScaffoldXXX"
                #scaffold = line_list[0].split('-')[1]    # if vsearch
                pos = line_list[1]
                snv_id = ':'.join([scaffold, pos])
                af = re.findall('AF=[0.0-9.0a-z\-]+', line_list[7])
                af = float(af[0].split('=')[1])
                rf = 1 - af
                af_line = str(snv_id + ' ' + str(af) + '\n')
                af_file.write(af_line)
                rf_line = str(snv_id + ' ' + str(rf) + '\n')
                rf_file.write(rf_line)
                for j, ind in enumerate(ind_ids):
                    if line_list[j+9][0:3] == './.':
                        rr, ra, aa = map(str,[0, 0, 0])
                        ind_line = str(rr + ' ' + ra + ' ' + aa)
                        n_no_num += 1
                        geno_likely[ind] = ind_line
                    else:
                        n_num += 1
                        gt_lklhd = line_list[j+9].split(':')
                        lklhd = gt_lklhd[4].split('\n')[0]
                        rr, ra, aa = lklhd.split(',')
                        ind_line = str(rr + ' ' + ra + ' ' + aa)
                        geno_likely[ind] = ind_line
                    line_dict[snv_id] = ' '.join(geno_likely.values())
    file.close()
    af_file.close()
    rf_file.close()


print n_ind, n_snv #, n_no_num, n_num
print ' '.join(map(str, ind_ids))
it = iter(sorted(line_dict.iteritems()))
for key, value in it:
    print key, value

File: scripts/vcf2gl.pl
#!/usr/bin/perl
#
# This scripts conversts a vcf file to a simpler format for downstream analysis. I am calling this format genetoype likelihood (gl). The first line lists: the number of individuals and loci. The next line has individual ids. This if followed by on line per SNP that gives the SNP id (scaffold, position) and the phred-scaled genotype likelihoods, three per individual. The script prints the non-reference allele frequency to a file so that they can be used to estimate genotypes
#
# USAGE: perl vcf2gl.pl myfile.vcf
#

my $in = shift(@ARGV);
my @line = ();
my $word;
my $nind = 0;
my $nloc = 0;
my $out = $in;
$out =~ s/vcf/gl/ or die "failed file name sub\n";

open (OUT, "> $out") or die "Could not write the outfile\n";

if ($out =~ s/gl/txt/){
	open (OUT2, "> af_$out") or die "Count not write the alt. af file\n";
}

open (IN, $in) or die "Could not read the vcf file\n";
while (<IN>){
	chomp;
	## get individual ids
	if (m/^#CHROM/){
		@line = split(m/\s+/, $_);	
		foreach $word (@line){
			if ($word =~ m/[0-9a-zA-Z_]*/){ ## this is an individual id
				push (@inds, $word);
				$nind++;
			}
		}
		## NOTE, number of loci is not correct, need to add manually after the fact
		## somdeday I will fix this
		print OUT "$nind $nloc\n";
		$word = join (" ", @inds);
		print OUT "$word\n";
	}
	## read genetic data lines, write gl
	elsif (m/^contig_(\d+)\s+(\d+)/){# modify this regular expression as needed
			$word = "$1".":"."$2";
			if (m/AC=(\d+)/){
				$p = $1;
				$p = $p/($nind*2.0);
				#print "$word = $p\n";
				$nloc++;
				print OUT "$word ";
				@line = split(m/\s+/, $_);
				$i = 0;
				foreach $word (@line){
					if ($word =~ s/^\d\/\d\://){
						$word =~ s/\d+,\d+:\d+:\d+:// or die "failed sub $word \n";
						$word =~ s/,/ /g;
						print OUT " $word";
					}
					elsif ($word =~ m/\.\/\./){
						print OUT " 0 0 0";
					}
			
				}
				print OUT "\n";
				print OUT2 "$p\n"; ## print p before converting to maf
			}
		}
	}	

close (OUT);
close (OUT2);
print "Number of loci: $nloc; number of individuals $nind\n";
